File: backend/src/main/java/com/backend/ord/config/security/JwtService.java
package com.backend.ord.config.security;

import com.backend.ord.config.properties.JwtProperties;
import com.backend.ord.domain.entities.User;
import com.backend.ord.services.UserService;
import com.backend.ord.utils.CookieUtils;
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.ExpiredJwtException;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import io.jsonwebtoken.io.Decoders;
import io.jsonwebtoken.security.Keys;
import jakarta.servlet.http.HttpServletRequest;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Service;

import java.security.Key;
import java.util.*;
import java.util.function.Function;

@Service
@RequiredArgsConstructor
public class JwtService {
    private final UserService userService;
    private final JwtProperties jwtProperties;

    public Optional<String> getJWTFromRequest(HttpServletRequest request) {
        return CookieUtils.getCookieValue(
                jwtProperties.getAuthCookieName(),
                request
        );
    }

    public Optional<User> getAuthenticatedUser(HttpServletRequest request) {
        Optional<String> jwtToken = getJWTFromRequest(request);

        if (jwtToken.isEmpty()) return Optional.empty();

        return userService.findUserByAuthToken(jwtToken.get());
    }

    public String extractUsername(String jwtToken) throws ExpiredJwtException {
        return extractClaim(jwtToken, Claims::getSubject);
    }

    public UUID extractUserId(String jwtToken) {
        String claimName = jwtProperties.getUserIdClaimName();

        return UUID.fromString(
                extractClaim(jwtToken, claims -> claims.get(claimName, String.class))
        );
    }

    public <T> T extractClaim(String jwtToken, Function<Claims, T> claimsResolver) {
        final Claims claims = extractAllClaims(jwtToken);
        return claimsResolver.apply(claims);
    }

    public Date extractExpiration(String jwtToken) {
        return extractClaim(jwtToken, Claims::getExpiration);
    }

    public String generateToken(UserDetails userDetails) {
        return generateToken(new HashMap<>(), userDetails);
    }

    public String generateToken(
            Map<String, Object> extraClaims,
            UserDetails userDetails
    ) {
        // Current time in milliseconds
        final long currentTime = System.currentTimeMillis();

        long JWT_EXPIRATION_TIME = jwtProperties.getExpirationTime();

        return Jwts.builder()
                .setClaims(extraClaims)
                .setSubject(userDetails.getUsername())
                .setIssuedAt(new Date(currentTime))
                .setExpiration(new Date(currentTime + JWT_EXPIRATION_TIME))
                .signWith(getSigningKey(), SignatureAlgorithm.HS256)
                .compact();
    }


    // ----------------
    // ### Private helper methods

    private Key getSigningKey() {
        String JWT_SECRET = jwtProperties.getSecretKey();

        byte[] keyBytes = Decoders.BASE64.decode(JWT_SECRET);
        return Keys.hmacShaKeyFor(keyBytes);
    }


    private Claims extractAllClaims(String jwtToken) {
        return Jwts.parserBuilder()
                .setSigningKey(getSigningKey())
                .build()
                .parseClaimsJws(jwtToken)
                .getBody();
    }
}

File: backend/src/main/kotlin/com/backend/ord/config/security/JwtService.kt
package com.backend.ord.config.security

import com.backend.ord.config.properties.JwtProperties
import com.backend.ord.domain.persistance.entities.User
import com.backend.ord.exceptions.REST.ForbiddenException
import com.backend.ord.services.UserService
import com.backend.ord.utils.CookieUtils.getCookieValue
import io.jsonwebtoken.Claims
import io.jsonwebtoken.ExpiredJwtException
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import io.jsonwebtoken.io.Decoders
import io.jsonwebtoken.security.Keys
import jakarta.servlet.http.HttpServletRequest
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.stereotype.Service
import java.security.Key
import java.util.*

@Service
class JwtService(
    private val userService: UserService,
    private val jwtProperties: JwtProperties
) {
    fun getJWTFromRequest(request: HttpServletRequest): String? =
        getCookieValue(jwtProperties.authCookieName, request)

    fun getAuthenticatedUser(request: HttpServletRequest): User? =
        getJWTFromRequest(request)?.let { jwtToken ->
            val userId = extractUserId(jwtToken)
            userService.findById(userId)
        }

    @Throws(ForbiddenException::class)
    fun getAuthenticatedUserOrThrowForbidden(request: HttpServletRequest): User =
        getAuthenticatedUser(request) ?: throw ForbiddenException("User is not authenticated")

    @Throws(ExpiredJwtException::class)
    fun extractUsername(jwtToken: String): String =
        extractClaim(jwtToken) { it.subject }

    fun extractUserId(jwtToken: String): UUID {
        val claimName = jwtProperties.userIdClaimName
        return UUID.fromString(
            extractClaim(jwtToken) { claims -> claims[claimName, String::class.java] }
        )
    }

    fun <T> extractClaim(jwtToken: String, claimsResolver: (Claims) -> T): T {
        val claims = extractAllClaims(jwtToken)
        return claimsResolver(claims)
    }

    fun extractExpiration(jwtToken: String): Date =
        extractClaim(jwtToken) { it.expiration }

    fun generateToken(userDetails: UserDetails): String =
        generateToken(emptyMap(), userDetails)

    fun generateToken(
        extraClaims: Map<String, Any>,
        userDetails: UserDetails
    ): String {
        val currentTime = System.currentTimeMillis()
        val expirationTime = jwtProperties.expirationTime

        return Jwts.builder()
            .setClaims(extraClaims)
            .setSubject(userDetails.username)
            .setIssuedAt(Date(currentTime))
            .setExpiration(Date(currentTime + expirationTime))
            .signWith(signingKey, SignatureAlgorithm.HS256)
            .compact()
    }

    private val signingKey: Key
        get() {
            val keyBytes = Decoders.BASE64.decode(jwtProperties.secretKey)
            return Keys.hmacShaKeyFor(keyBytes)
        }

    private fun extractAllClaims(jwtToken: String): Claims =
        Jwts.parserBuilder()
            .setSigningKey(signingKey)
            .build()
            .parseClaimsJws(jwtToken)
            .body
}
